### PR TITLE
oneDAL kmeans initialization algorithm

### PR DESCRIPTION
**Pull request by mchernov-intel**
oneDAL kmeans initialization algorithm
kmeans_init

### CODE REVIEW

File: source/elements/oneDAL/source/algorithms/clustering/kmeans.rst [hunk 1]
```diff
@@ -136,23 +136,23 @@ Descriptor
...
+      descriptor();
 
       int64_t get_cluster_count() const;
       int64_t get_max_iteration_count() const;
       double get_accuracy_threshold() const;
 
-      desc& set_cluster_count(int64_t);
-      desc& set_max_iteration_count(int64_t);
-      desc& set_accuracy_threshold(double);
+      descriptor& set_cluster_count(int64_t);
```
[the remark this answers is not in the document]
Comment (mchernov-intel) [reply source/elements/oneDAL/source/algorithms/clustering/kmeans.rst]: decided @rlnx will fix it in another pull request

File: source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+-------------
+Usage example
+-------------
+::
+
+   onedal::table run_compute(const onedal::table& data) {
+      const auto kmeans_desc = onedal::kmeans_init::descriptor<float>{}
+         .set_cluster_count(10)
+         .set_max_iteration_count(50)
```
[the remark this answers is not in the document]
Comment (mchernov-intel) [reply source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst]: updated

File: source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+Given the training set :math:`X = \{ x_1, \ldots, x_n \}` of
+:math:`p`-dimensional feature vectors and a positive integer :math:`k`, the
+problem is to find a set :math:`C = \{ c_1, \ldots, c_k \}` of
+:math:`p`-dimensional initial centroids.
+
+--------------
+Default dense
+--------------
+
+Choice of first :math:`k` feature vectors from the training set :math:`X`
```
[the remark this answers is not in the document]
Comment (mchernov-intel) [reply source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst]: fixed

File: source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst [hunk 4]
```diff
@@ -0,0 +1,41 @@
+.. highlight:: cpp
+.. default-domain:: cpp
+
+======================
+K-Means initialization
+======================
+
+The K-Means initialization algorithm gets :math:`n` feature vectors as input and
+chooses :math:`k` initial centroids. After initialization, K-Means algorithm
+could be run to partition input data into :math:`k` clusters.
```
[the remark this answers is not in the document]
Comment (mchernov-intel) [reply source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst]: fixed

**Comment by mchernov-intel** [comment]
@rlnx @andreynv @outoftardis please, review

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+///                intermediate computations. Can be :expr:`float` or
+///                :expr:`double`.
+/// @tparam Method – Tag-type that specifies an implementation
+/// of K-Means Initialization algorithm.
+template <typename Float = float,
+          typename Method = method::by_default>
+class descriptor {
+public:
+
+   /// Creates new instance of descriptor with the default attribute values.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:23]: "a new instance of the descriptor"

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+   /// Creates input for the computing operation with the default attribute values.
+   /// @remark default = table{}
+   compute_input();
+
+   /// Creates input for the computing operation with the given data,
+   /// the other attributes get default values.
+   compute_input(const table& data = table{});
+
+   /// The computing set $X$
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:44]: Not clear to me what this 'computing set' is. The set that the computations are performed on?

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+public:
+
+   compute_result();
+
+   /// The initial centroids
+   /// @remark default = table{}
+   table get_centroids() const;
+};
+
+/// Runs the computing operation for KMeans initialization. For more details see
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:60]: Comma after details

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+/// :expr:`oneapi::dal::compute`.
+///
+/// @tparam Float  The floating-point type that the algorithm uses for
+///                intermediate computations. Can be :expr:`float` or
+///                :expr:`double`.
+/// @tparam Method Tag-type that specifies an implementation
+/// of K-Means Initialization algorithm.
+///
+/// @param[in] desc  Descriptor of the algorithm
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:69]: The descriptor

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+/// @param[in] input Input data for the training operation
+/// @return result Result of the training operation
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:71]: The result

File: source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst [hunk 10]
```diff
@@ -0,0 +1,62 @@
+.. highlight:: cpp
+.. default-domain:: cpp
+
+======================
+K-Means initialization
+======================
+
+The K-Means initialization algorithm gets :math:`n` feature vectors as input and
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst:8]: Gets -> receives

File: source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+Given the training set :math:`X = \{ x_1, \ldots, x_n \}` of
+:math:`p`-dimensional feature vectors and a positive integer :math:`k`, the
+problem is to find a set :math:`C = \{ c_1, \ldots, c_k \}` of
+:math:`p`-dimensional initial centroids.
+
+-------------
+Default dense
+-------------
+
+Method choses first :math:`k` feature vectors from the training set :math:`X`
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst:21]: The method
Full stop in the end of the sentence

File: source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst [hunk 12]
```diff
@@ -0,0 +1,62 @@
+.. highlight:: cpp
+.. default-domain:: cpp
+
+======================
+K-Means initialization
+======================
+
+The K-Means initialization algorithm gets :math:`n` feature vectors as input and
+chooses :math:`k` initial centroids. After initialization, K-Means algorithm
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst:9]: Don't you want to mention how it chooses centroids? Randomly or not?

Comment (mchernov-intel) [reply source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst:9]: Depends on method, so. no

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp (hunk 6 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp]: fixed

**Comment by mchernov-intel** [comment]
@rlnx @andreynv @outoftardis please, review

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+   /// computational method.
+   struct dense {};
+
+   using by_default = dense;
+} // namespace method
+
+/// @tparam Float  The floating-point type that the algorithm uses for
+///                intermediate computations. Can be :expr:`float` or
+///                :expr:`double`.
+/// @tparam Method – Tag-type that specifies an implementation
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:17]:
```suggestion
/// @tparam Method Tag-type that specifies an implementation
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+   /// Creates a new instance of the class with the default property values.
+   compute_result();
+
+   /// The initial centroids
+   /// @remark default = table{}
+   const table& get_centroids() const;
+};
+
+/// Runs the computing operation for KMeans initialization. For more details,
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:60]:
```suggestion
/// Runs the computing operation for K-Means initialization. For more details,
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+/// see :expr:`oneapi::dal::compute`.
+///
+/// @tparam Float  The floating-point type that the algorithm uses for
+/// intermediate computations. Can be :expr:`float` or :expr:`double`. @tparam
+/// Method Tag-type that specifies an implementation of K-Means Initialization
+/// algorithm.
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:66]: can you make it more readable and split the descriptions of the parameters into separate lines?

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+///
+/// @param[in] desc  The descriptor of the algorithm @param[in] input Input data
+/// for the computing operation @return result The result of the computing
+/// operation
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:70]: same comment about readability

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+///
+/// @pre  :expr:`input.data.has_data == true` @pre  :expr:`input.data.row_count
+/// == desc.cluster_count` @post :expr:`result.centroids.has_data == true` @post
+/// :expr:`result.centroids.row_count == desc.cluster_count` @post
+/// :expr:`result.centroids.column_count == input.data.column_count`
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:75]: same comment about readability

File: source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+:math:`p`-dimensional feature vectors and a positive integer :math:`k`, the
+problem is to find a set :math:`C = \{ c_1, \ldots, c_k \}` of
+:math:`p`-dimensional initial centroids.
+
+.. _kmeans_init_c_math_dense:
+
+Computing method: *dense*
+-------------------------
+
+The method choses first :math:`k` feature vectors from the training set
```
Comment (outoftardis) [inline source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst:43]:
```suggestion
The method chooses the first :math:`k` feature vectors from the training set
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+class compute_result {
+public:
+
+   /// Creates a new instance of the class with the default property values.
+   compute_result();
+
+   /// The initial centroids
+   /// @remark default = table{}
+   const table& get_centroids() const;
```
Comment (andreynv) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:57]: We have to specify layout of the table-how the tables stores centroids - use k-means description as the example

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp (hunk 13 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp]: fixed

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp (hunk 14 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp]: fixed

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp (hunk 15 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp]: fixed

File: source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst (hunk 18 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/source/algorithms/clustering/kmeans_init.rst]: fixed

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp (hunk 19 above)
Comment (mchernov-intel) [reply source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:59]: fixed also for input too.

**Comment by mchernov-intel** [comment]
@rlnx @andreynv @outoftardis please, review

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+};
+
+class compute_input {
+public:
+
+   /// Creates input for the computing operation with the given data,
+   /// the other attributes get default values.
+   compute_input(const table& data = table{});
+
+   /// $n \\times p$ table with the data to be clustered, where each row
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:43]:
```suggestion
   /// An $n \\times p$ table with the data to be clustered, where each row
```

File: source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+   compute_input& set_data(const table&);
+};
+
+class compute_result {
+public:
+
+   /// Creates a new instance of the class with the default property values.
+   compute_result();
+
+   /// $k \\times p$ table with the initial centroids. Each row of the
```
Comment (outoftardis) [inline source/elements/oneDAL/include/oneapi/dal/kmeans_init.hpp:56]:
```suggestion
   /// A $k \\times p$ table with the initial centroids. Each row of the
```

**Comment by andreynv** [comment]
I do not have other comments, if Katya has comments for gramma they should be fixed

**Review by andreynv — APPROVED** [review]

**Comment by outoftardis** [comment]
no more comments from me

**Pull request merged**